### ISSUE TITLE
Unclear semantics if there are multiple `NXsample` groups

### ISSUE THREAD

**Issue by paulmillar** [nexusformat/definitions#1361]
The current definition of `NXsample` is:

```xml
<doc>
    Any information on the sample. 

    This could include scanned variables that
    are associated with one of the data dimensions, e.g. the magnetic field, or
    logged data, e.g. monitored temperature vs elapsed time.
</doc>
```

This definition doesn't make it clear what is the semantic relationship between multiple `NXsample` groups.

In more concrete terms...

It's fairly clear that a single `NXsample` group describes a single sample (the above description says "the sample").

However, if a `NXentry` contains multiple `NXsample` groups, does each `NXsample` group describe the same sample or potentially different samples?

If a NeXus file contains multiple `NXentry` groups, each with a single `NXsample` group, are these `NXsample` group the same sample or potentially different samples?

**Comment by prjemian** [comment]
> If a NeXus file contains multiple NXentry groups, each with a single NXsample group, are these NXsample group the same sample or potentially different samples?

Generally, they are different samples, defined only within their parent NXentry group.  If you want to make them the same, write the same information or use HDF5 links.

**Comment by prjemian** [comment]
> ... if a NXentry contains multiple NXsample groups ...

Can you describe a specific use case for such?

**Comment by prjemian** [comment]
Without a `maxOccurs="1"` attribute in the declaration of `NXsample` within `NXentry`: https://github.com/nexusformat/definitions/blob/4c09c7718c41dc90eb996475efdf1c0d30fb1d5d/base_classes/NXentry.nxdl.xml#L223

then the default is `"unbounded"`: https://github.com/nexusformat/definitions/blob/4c09c7718c41dc90eb996475efdf1c0d30fb1d5d/nxdl.xsd#L412

This means that multiple NXsample groups are allowed.

When you provide multiple NXsample groups, it's up to you to define if they are the same or different **and** to describe how the multiple groups should be handled by users of such data.

**Comment by paulmillar** [comment]
The specific use case that has come up is describing the sample environment.  This comes from the work on embedding SECoP information in NeXus.

The current proposed strategy involves creating two `NXSample` groups: one with `type` field of `sample` and the other with the `type` field of `sample environment`.  Both `NXSample` groups are placed under the `NXentry`.

(I believe there are [some issues](https://github.com/nexusformat/definitions/issues/1366) with this `type` field, but I think that is largely independent from this issue.)

You can see example files that demonstrate this in [this repo](https://github.com/jkotan/secop-file-examples), and can (interactively) view those file's contents using myHDF5: [sample_env_00591.nxs](https://myhdf5.hdfgroup.org/view?url=https%3A%2F%2Fgithub.com%2Fjkotan%2Fsecop-file-examples%2Fblob%2Fmain%2Fsample_env_00591.nxs) and [sample_env_00592.nxs](https://myhdf5.hdfgroup.org/view?url=https%3A%2F%2Fgithub.com%2Fjkotan%2Fsecop-file-examples%2Fblob%2Fmain%2Fsample_env_00592.nxs).

**Comment by paulmillar** [comment]
> This means that multiple NXsample groups are allowed.

Indeed.  That was my conclusion, too.

> When you provide multiple NXsample groups, it's up to you to define if they are the same or different and to describe how the multiple groups should be handled by users of such data.

Well, OK.

But then, my follow-on question would be ... how?

If NeXus has this flexibility (i.e., there are no restrictions placed on the relationship between multiple `NXsample` groups) then how should the person/agent writing the NeXus file indicate the intended relationship between multiple `NXsample` groups?

This could be described in the `description` field.  However, I believe that field is intended for human consumption.

**Comment by prjemian** [comment]
> The current proposed strategy involves creating two `NXSample` groups: one with `type` field of `sample` and the other with the `type` field of `sample environment`. Both `NXSample` groups are placed under the `NXentry`.

Thanks for that description.  As shown above, multiple `NXsample` groups are allowed by the NXDL language.

As you noted, a custom attribute of `type` is problematic due to conflict with other use in the NXDL language.  (In fields, `type` refers to the storage type, such as `NX_CHAR`.  In groups, `type` refers to the name of the NeXus class definition, such as `NXuser`.)  Pick a different

**Comment by prjemian** [comment]
> But then, my follow-on question would be ... how?

AFAIK, NeXus has no formal mechanism.  I recall discussions (about `maxOccurs`) but the assumption to date has not involved multiple NXsample groups.  Make a proposition showing how to distinguish the relationship.  

Could the sample environment be an `environment` (`type="NXenvironment"`) group within NXsample?

**Comment by paulmillar** [comment]
I agree that `type` (as an attribute within the `NXsample` class) is problematic.  Please see #1366 for more comments on this theme. (I've referenced your comment there, so it isn't lost.)

However, please note that this `type` attribute (within `NXsample`) in not custom, but an already established part of NeXus.  In fact, the attribute was part of the initial definition of `NXsample` and (seemingly) has not modified since.

**Comment by prjemian** [comment]
The use of `type` in `NXsample` is a `field`, not an `attribute`: https://github.com/nexusformat/definitions/blob/4c09c7718c41dc90eb996475efdf1c0d30fb1d5d/base_classes/NXsample.nxdl.xml#L204-L217

The name is confusing here, yet historical (and tough to be changed).  A lesson learned.  Perhaps.

**Comment by paulmillar** [comment]
> Could the sample environment be an environment (type="NXenvironment") group within NXsample?

I was thinking along similar lines, too.

However, although not explicitly documented  (see #1362), it seems that `NXenvironment` groups are intended to describe individual specific apparatuses/devices.

It seems the current proposed way of including SECoP information in NeXus takes this interpretation.

It has multiple `NXenvironment` groups, each located within the same `NXsample` / `type=sample environment` group.  These `NXenvironment` groups each describe a different physical device (cryo, magnets, 3He compressor/value, etc) that have been identified and monitored through SECoP.

**Comment by paulmillar** [comment]
> The use of type in NXsample is a field, not an attribute

Sorry, you're right: `type` is a field and not an attribute.
